File: search/src/main/scala/weco/api/search/elasticsearch/ResilientElasticClient.scala
package weco.api.search.elasticsearch

import com.sksamuel.elastic4s.{ElasticClient, ElasticRequest, Handler, Response}
import grizzled.slf4j.Logging

import java.time.Clock
import java.util.concurrent.{Executors, ThreadFactory, TimeUnit}
import scala.concurrent.{ExecutionContext, Future, Promise}
import scala.util.control.NonFatal
import scala.util.{Failure, Success}

class ResilientElasticClient(
  clientFactory: () => ElasticClient,
  minRefreshIntervalMs: Long = 2000,
  closeGraceMs: Long = 30000,
  probeTimeoutMs: Long = 5000
)(implicit clock: Clock, ec: ExecutionContext)
    extends Logging
    with AutoCloseable {

  @volatile private var client: ElasticClient = clientFactory()
  @volatile private var lastRefreshTime: Long = 0

  // Dedicated thread keeps the factory's blocking Secrets Manager IO off the request dispatcher
  private val refreshExecutor = Executors.newSingleThreadScheduledExecutor(
    new ThreadFactory {
      override def newThread(r: Runnable): Thread = {
        val thread = new Thread(r, "resilient-elastic-client-refresh")
        thread.setDaemon(true)
        thread
      }
    }
  )
  private val refreshEc = ExecutionContext.fromExecutor(refreshExecutor)

  private val refreshLock = new Object
  private var inFlightRefresh: Option[Future[Unit]] = None
  private var pendingClose: Option[ElasticClient] = None

  def execute[T, U](t: T)(implicit handler: Handler[T, U],
                          manifest: Manifest[U]): Future[Response[U]] = {
    val currentClient = client
    currentClient.execute(t).transformWith {
      case Success(response)
          if response.status == 401 || response.status == 403 =>
        warn(
          s"Received ${response.status} from Elasticsearch, refreshing client and retrying...")
        retryIfRefreshed(currentClient, t, response)
      case Success(response)                               => Future.successful(response)
      case Failure(NonFatal(e)) if client ne currentClient =>
        // Our client was replaced mid-flight and likely closed under us; retry
        // through execute so a 401 on the retry still gets refresh handling.
        // Re-entry is bounded: each level requires another swap to have happened.
        warn(
          "Request failed on a replaced Elasticsearch client, retrying on the current one",
          e)
        execute(t)
      // Transport errors are not rotation evidence: fail fast, no refresh
      case Failure(e) => Future.failed(e)
    }
  }

  // Join the coalesced refresh, then retry only if the client actually changed:
  // a retry on the same client is futile and just duplicates load
  private def retryIfRefreshed[T, U](failedClient: ElasticClient,
                                     t: T,
                                     originalResponse: Response[U])(
    implicit handler: Handler[T, U],
    manifest: Manifest[U]): Future[Response[U]] =
    triggerRefresh(failedClient).flatMap { _ =>
      val current = client
      if (current ne failedClient) current.execute(t)
      else Future.successful(originalResponse)
    }

  // Coalesces concurrent requests onto one in-flight refresh instead of queueing on a lock
  private def triggerRefresh(failedClient: ElasticClient): Future[Unit] =
    refreshLock.synchronized {
      inFlightRefresh.getOrElse {
        if (client ne failedClient) {
          info("Elasticsearch client already refreshed by another thread.")
          Future.unit
        } else {
          val now = clock.millis()
          if (now - lastRefreshTime > minRefreshIntervalMs) {
            val refresh = doRefresh()
            inFlightRefresh = Some(refresh)
            refresh.onComplete { _ =>
              refreshLock.synchronized { inFlightRefresh = None }
            }(refreshEc)
            refresh
          } else {
            warn(
              s"Refresh requested too soon (last refresh ${now - lastRefreshTime}ms ago). " +
                s"Skipping, waiting on cooldown: ${minRefreshIntervalMs}ms"
            )
            Future.unit
          }
        }
      }
    }

  private def doRefresh(): Future[Unit] = {
    info("Refreshing Elasticsearch client...")
    Future(clientFactory())(refreshEc)
      .flatMap { candidate =>
        probeAccepts(candidate).map { accepted =>
          if (accepted) swapIn(candidate)
          else {
            // Secrets Manager may still be serving the invalidated key; keep the
            // current client and probe again on the next post-cooldown failure
            warn("Rebuilt Elasticsearch client failed to authenticate; keeping current client.")
            if (candidate ne client) closeQuietly(candidate)
          }
        }(refreshEc)
      }(refreshEc)
      .recover {
        case NonFatal(e) => error("Failed to refresh Elasticsearch client", e)
      }(refreshEc)
      // Throttle the next refresh attempt whether or not this one healed us
      .andThen { case _ => lastRefreshTime = clock.millis() }(refreshEc)
  }

  // Reject only a definitive 401/403: an inconclusive probe (5xx, 429, 404, timeout,
  // exception) must not block adopting a possibly-good key during a cluster incident,
  // and security-disabled clusters answer 4xx/404 here
  private def probeAccepts(candidate: ElasticClient): Future[Boolean] = {
    val promise = Promise[Boolean]()
    candidate.client.send(
      ElasticRequest("GET", "/_security/_authenticate"), {
        case Right(response) =>
          promise.trySuccess(
            response.statusCode != 401 && response.statusCode != 403)
        case Left(_) => promise.trySuccess(true)
      }
    )
    val timeout = refreshExecutor.schedule(new Runnable {
      override def run(): Unit = promise.trySuccess(true)
    }, probeTimeoutMs, TimeUnit.MILLISECONDS)
    // Don't leave a timer queued for every probe that answered promptly
    promise.future.andThen { case _ => timeout.cancel(false) }(refreshEc)
  }

  private def swapIn(candidate: ElasticClient): Unit = {
    val oldClient = client
    client = candidate
    // The fixture factory can return the same client, so never close it in that case
    if (oldClient ne candidate) deferClose(oldClient)
    info("Elasticsearch client refreshed.")
  }

  // Grace period lets in-flight requests on the old client complete; at most one
  // close is deferred, bounding connection-pool buildup if swaps repeat
  private def deferClose(oldClient: ElasticClient): Unit =
    refreshLock.synchronized {
      pendingClose.foreach(closeQuietly)
      pendingClose = Some(oldClient)
      refreshExecutor.schedule(
        new Runnable {
          override def run(): Unit =
            refreshLock.synchronized {
              if (pendingClose.exists(_ eq oldClient)) {
                pendingClose = None
                closeQuietly(oldClient)
              }
            }
        },
        closeGraceMs,
        TimeUnit.MILLISECONDS
      )
    }

  private def closeQuietly(c: ElasticClient): Unit =
    try c.close()
    catch {
      case NonFatal(e) => warn("Failed to close Elasticsearch client", e)
    }

  override def close(): Unit = {
    refreshExecutor.shutdownNow()
    refreshLock.synchronized {
      pendingClose.foreach(closeQuietly)
      pendingClose = None
    }
    closeQuietly(client)
  }
}

File: search_v2/src/services/resilient-elastic-client.ts
import { Client } from "@elastic/elasticsearch";
import {
  SearchRequest,
  SearchResponse,
  GetRequest,
  GetResponse,
  ClusterHealthRequest,
  ClusterHealthResponse,
} from "@elastic/elasticsearch/lib/api/types";
import log from "./logging";

type ClientFactory = () => Promise<Client>;

/**
 * A wrapper around the Elasticsearch client that handles credential expiry.
 *
 * When the client receives a 401 or 403 response from Elasticsearch, it will:
 * 1. Refresh the client by calling the factory function to get new credentials
 * 2. Retry the failed request with the new client
 *
 * Features:
 * - Auto-retry on 401/403 responses
 * - Cooldown throttling to prevent excessive credential refreshes
 * - Thread-safe refresh (only one refresh at a time via promise coordination)
 */
export class ResilientElasticClient {
  private client: Client;
  private clientFactory: ClientFactory;
  private minRefreshIntervalMs: number;
  private lastRefreshTime: number = 0;
  private refreshPromise: Promise<void> | null = null;

  constructor(
    initialClient: Client,
    clientFactory: ClientFactory,
    minRefreshIntervalMs: number = 2000
  ) {
    this.client = initialClient;
    this.clientFactory = clientFactory;
    this.minRefreshIntervalMs = minRefreshIntervalMs;
  }

  /**
   * Get the underlying Elasticsearch client.
   * This is needed for direct access to client methods like search, get, etc.
   */
  getClient(): Client {
    return this.client;
  }

  /**
   * Execute a request with automatic retry on auth failures.
   *
   * @param operation - A function that performs the ES operation using the client
   * @param getStatusCode - A function to extract status code from the response (for meta responses)
   * @returns The result of the operation
   */
  private async executeWithRetry<T>(
    operation: (client: Client) => Promise<T>,
    getStatusCode?: (response: T) => number | undefined
  ): Promise<T> {
    const currentClient = this.client;

    try {
      const response = await operation(currentClient);

      // Check for auth failures that need retry (if we can get status code)
      const statusCode = getStatusCode?.(response);
      if (statusCode === 401 || statusCode === 403) {
        log.warn(
          `Received ${statusCode} from Elasticsearch, refreshing client and retrying...`
        );
        await this.refreshClient(currentClient);
        return operation(this.client);
      }

      return response;
    } catch (error: unknown) {
      // Handle errors that might indicate auth issues
      if (this.isAuthError(error)) {
        log.warn(
          "Received auth error from Elasticsearch, refreshing client and retrying..."
        );
        await this.refreshClient(currentClient);
        return operation(this.client);
      }
      throw error;
    }
  }

  /**
   * Perform a search operation with automatic retry on auth failures.
   */
  async search<TDocument = unknown>(
    params: SearchRequest
  ): Promise<SearchResponse<TDocument>> {
    return this.executeWithRetry((client) => client.search<TDocument>(params));
  }

  /**
   * Perform a get operation with automatic retry on auth failures.
   */
  async get<TDocument = unknown>(
    params: GetRequest
  ): Promise<GetResponse<TDocument>> {
    return this.executeWithRetry((client) => client.get<TDocument>(params));
  }

  /**
   * Check cluster health with automatic retry on auth failures.
   */
  async clusterHealth(
    params?: ClusterHealthRequest
  ): Promise<ClusterHealthResponse> {
    return this.executeWithRetry((client) => client.cluster.health(params));
  }

  private isAuthError(error: unknown): boolean {
    if (error && typeof error === "object") {
      const statusCode = (error as { statusCode?: number }).statusCode;
      if (statusCode === 401 || statusCode === 403) {
        return true;
      }
      // Also check for meta.statusCode in ResponseError
      const meta = (error as { meta?: { statusCode?: number } }).meta;
      if (meta?.statusCode === 401 || meta?.statusCode === 403) {
        return true;
      }
    }
    return false;
  }

  private async refreshClient(failedClient: Client): Promise<void> {
    // If the client has already been refreshed by another request, don't refresh again
    if (this.client !== failedClient) {
      log.info("Elasticsearch client already refreshed by another request.");
      return;
    }

    // If a refresh is already in progress, wait for it
    if (this.refreshPromise) {
      log.info("Waiting for ongoing client refresh...");
      await this.refreshPromise;
      return;
    }

    // Check cooldown period
    const now = Date.now();
    if (now - this.lastRefreshTime < this.minRefreshIntervalMs) {
      log.warn(
        `Refresh requested too soon (last refresh ${
          now - this.lastRefreshTime
        }ms ago). ` +
          `Skipping, waiting on cooldown: ${this.minRefreshIntervalMs}ms`
      );
      return;
    }

    // Start the refresh process
    this.refreshPromise = this.doRefresh();
    try {
      await this.refreshPromise;
    } finally {
      this.refreshPromise = null;
    }
  }

  private async doRefresh(): Promise<void> {
    const oldClient = this.client;
    try {
      log.info("Refreshing Elasticsearch client...");
      this.client = await this.clientFactory();
      this.lastRefreshTime = Date.now();
      await oldClient.close();
      log.info("Elasticsearch client refreshed.");
    } catch (error) {
      log.error("Failed to refresh Elasticsearch client", error);
      throw error;
    }
  }
}

/**
 * Create a resilient Elasticsearch client with automatic retry on auth failures.
 */
export const createResilientClient = async (
  clientFactory: ClientFactory,
  minRefreshIntervalMs: number = 2000
): Promise<ResilientElasticClient> => {
  const initialClient = await clientFactory();
  return new ResilientElasticClient(
    initialClient,
    clientFactory,
    minRefreshIntervalMs
  );
};
